superset-frontend/src/explore/components/controls/AnnotationLayerControl/AnnotationLayer.jsx

```diff
@@ -298,74 +306,142 @@ class AnnotationLayer extends React.PureComponent {
     });
   }
 
-  fetchOptions(annotationType, sourceType, isLoadingOptions) {
-    if (isLoadingOptions) {
-      if (sourceType === ANNOTATION_SOURCE_TYPES.NATIVE) {
-        const queryParams = rison.encode({
-          page: 0,
-          page_size: FILTER_OPTIONS_LIMIT,
-        });
-        SupersetClient.get({
-          endpoint: `/api/v1/annotation_layer/?q=${queryParams}`,
-        }).then(({ json }) => {
-          const layers = json
-            ? json.result.map(layer => ({
-                value: layer.id,
-                label: layer.name,
-              }))
-            : [];
-          this.setState({
-            isLoadingOptions: false,
-            valueOptions: layers,
-          });
-        });
-      } else if (requiresQuery(sourceType)) {
-        const queryParams = rison.encode({
-          filters: [
-            {
-              col: 'id',
-              opr: 'chart_owned_created_favored_by_me',
-              value: true,
+  handleTextValue(inputValue) {
+    this.setState({
+      value: inputValue,
+    });
+  }
+
+  asyncFetch = async (_, page, pageSize) => {
+    const { annotationType, sourceType } = this.state;
+
+    if (sourceType === ANNOTATION_SOURCE_TYPES.NATIVE) {
+      const queryParams = rison.encode({
+        page,
+        page_size: pageSize,
+      });
+
+      const { json } = await SupersetClient.get({
+        endpoint: `/api/v1/annotation_layer/?q=${queryParams}`,
+      });
+
+      const { result, count } = json;
+
+      const layersObj = result.reduce((acc, layer) => {
+        acc[layer.id] = {
+          value: layer.id,
+          label: layer.name,
+        };
+        return acc;
+      }, {});
+
+      const layersArray = Object.values(layersObj);
+
+      this.setState(prevState => ({
+        valueOptions: { ...prevState.valueOptions, ...layersObj },
+      }));
+
+      return {
+        data: layersArray,
+        totalCount: count,
+      };
+    }
+
+    if (requiresQuery(sourceType)) {
```

Comment:
It would be a good idea to break this function into 2: one for `sourceType === ANNOTATION_SOURCE_TYPES.NATIVE` and another for `requiresQuery(sourceType)`

Reply:
hopefully I addressed this correctly